Print function names in backtraces

stedolan opened the pull request:
This PR adds function names to OCaml backtraces. As an example, here's what I see when I add a 'assert false' randomly somewhere in the middle of ctype.ml:
```
Fatal error: exception File "typing/ctype.ml", line 728, characters 5-11: Assertion failed
Raised at Ctype.generalize_structure in file "typing/ctype.ml", line 728, characters 5-17
Called from Typecore.type_expect_ in file "typing/typecore.ml", line 2859, characters 6-29
Called from Builtin_attributes.warning_scope in file "parsing/builtin_attributes.ml", line 236, characters 14-18
Re-raised at Builtin_attributes.warning_scope in file "parsing/builtin_attributes.ml", line 241, characters 10-13
Called from Typecore.type_expect in file "typing/typecore.ml", line 2327, characters 4-156
Called from Typecore.type_expect_ in file "typing/typecore.ml", line 3305, characters 17-39
[...]
```

I have two motivations:

  1. I want function names in backtraces
  2. In Statmemprof, it's natural to group backtraces of allocations that share a suffix (statmemprof-emacs does this). But the backtraces have precise locations, so two callsites in the same function are distinct backtrace entries, making the grouping too fine-grained to be useful.

Spacetime has some code to deal with 2 (which lives in [spacetime_lib](https://github.com/lpw25/spacetime_lib)), which associates backtrace entries (i.e. callsites) with functions by parsing the ELF symbol table and looking up function names via DWARF (using [owee](https://github.com/let-def/owee)). This works, but with a couple of disadvantages: it's fairly complicated, ELF-specific, and exposes the user to unpalatable generated names. (It also doesn't help with 1).

Instead, this patch extends the existing debug info with function names, which are looked up using the same code that currently finds source locations.

## Names and scopes

In OCaml, not all functions have names, and not all names are for functions.

This PR assigns names by dividing the source file into "scopes", which may nest but not overlap. Locations are named according to their stack of enclosing scopes, and new scopes are introduced by:
  - named functions
  - anonymous functions
  - modules
  - values bindings in modules
  - classes
  - methods

(See the type `Lambda.scoped_location` for how these are represented)

As an example, here's a simple program and the backtrace it prints:
```ocaml
let f () =
  List.iter (fun _ -> assert false) [1;2;3];
  print_string "done!"
let () = f ()
```
```
Fatal error: exception Assert_failure("raise_iter.ml", 2, 22)
Raised at Raise_iter.f.(fun) in file "raise_iter.ml", line 2, characters 22-34
Called from Stdlib__list.iter in file "list.ml", line 110, characters 12-15
Called from Raise_iter.f in file "raise_iter.ml", line 2, characters 2-43
Called from Raise_iter in file "raise_iter.ml", line 4, characters 9-13
```

The names printed above (intentionally) look like module paths. However, these are not in general paths: for instance, some mention `(fun)`, and it is possible to have locations reported as e.g. `Foo.f.Bar.g` (by using `let module` to nest a module inside an expression), which is not a valid path.

For a more comprehensive example, see [names.ml in the testsuite](https://github.com/ocaml/ocaml/pull/9096/commits/dde9b1e471ba838be4aa2e8c6fcd911368acdd87#diff-80009260911204ed35eb2457779906cd)

Since patterns need not bind only a single name, it can be ambiguous what the name of a scope is. In such cases, this patch arbitrarily picks one of the names: `let (f, g) = assert false` will be reported as an exception in `f`. If no names are bound by a pattern, then no scope is introduced. So, if a module `M` contains `let () = assert false`, the exception is reported as simply coming from `M`.

## Binary size impact

In native code, the extra debug information makes ocamlopt.opt about 2.5% bigger than current trunk, and similar for Flambda. Strings are shared when they occur multiple times. (An earlier version of this patch included a rather complicated compression scheme for sharing prefixes. I can revisit it later if the debug info size becomes a problem).

Bytecode uses a simpler format, and the impact there is much larger (about 10%). I note that debug info is already very large on bytecode (it includes a typing environment for ocamldebug's benefit) and 
binary sizes are larger than native (bytecode ocamlopt is about 2x bigger than ocamlopt.opt). From this I conclude that nobody cares very much about bytecode binary size, so I did the simplest possible thing there and wrote strings directly. (More optimised versions are of course possible).

## Reviewing notes

This is a large patch, but I've arranged the commit history to make it easier to review. The commits are:

```
  1. 888ce57634 Introduce the Lambda.scoped_location type
  2. 541d829457 Replace Location.t with Lambda.scoped_location in Lambda code
  3. 07da0a18e4 Print function names (derived from Lambda.scoped_location) in backtraces
  4. b4550dd1b3 Generate function names during translation from Typedtree to Lambda
  5. dde9b1e471 Update testsuite with new backtrace format, and add test for names
  6. 88f157703a  Bootstrap
```
Commits 1, 3 and 4 contain the actual work, and require review. They're relatively small.

Commit 2 is very large but very simple. The only thing it does is thread a new `~scopes` parameter around all of the `transl_foo` functions that translate Typedtree to Lambda, so that the raw locations of Typedtree can be turned into `scoped_locations` in Lambda.

Commit 5 is also large, because it updates all of the backtraces in the testsuite. (Commit 6 is a bootstrap).

gasche commented:
> Note to reviewers: this patch looks large only because it's on top of the not-yet-merged #8805. There's only one new commit in this PR, currently at +275 -140 lines.

I solved that for you.

I haven't looked at the code. The description of the design is sensible and I agree it's a general feature. I would be interested in rephrasing the current wording of the backtrace lines, `Called from <function> at <location>`, to suggest less strongly that `<function>` is the caller, given the approximations introduced by the naming scheme. What about `Called within <function> at <location>`, for example?

stedolan commented:
> I solved that for you.

Thanks!

> I haven't looked at the code. The description of the design is sensible and I agree it's a general feature. I would be interested in rephrasing the current wording of the backtrace lines, `Called from <function> at <location>`, to suggest less strongly that `<function>` is the caller, given the approximations introduced by the naming scheme. What about `Called within <function> at <location>`, for example?

I'd claim that it's not an approximation, but that `Called from <foo> ...` means "called from within the definition of foo" rather than "called from within the body of the function named foo" :)

Pedantry aside, I'm not attached to the current wording (I didn't put much thought into it), but I'm not really seeing the same subtle distinction between "called from" and "called within" - they mean the same to me, but the latter sounds a bit more awkward.

The functions vs. scopes thing is a bit subtle, though (cf. the wall of text above), and I'm not really sure the best way to get it across here.

lpw25 commented:
I think that using the statically enclosing bindings like this is likely to be very confusing for users. It seems almost certain to be misread by people trying to track down a failure and to lead to wasted time looking in the wrong function.

I've tackled this issue before as part of work (not yet upstreamed) to improve flambda's inlining reports. I found that the nicest "name" for a function was along the lines of:
```
type parent_ =
  | Unit of string
  | Submodule of parent_ * string
  | Class of parent_ * string
  | Function of function_ * string

and function_name =
  | Name of string
  | Anonymous of Location.t

and function_ = parent_ * function_name
```
Which allows you to print things like: `the anonymous function at foo.ml:3,1-10 within Foo.bar` or `Bar.Baz.foo`.

That might be a bit too detailed for use in backtraces, but at the very least I think you can get them to distinguish: `an anonymous function within Foo.bar` from `Foo.bar`.

alainfrisch commented:
If we have a nice symbolic path to point to a location, it would be nice to include it in the payload for `Assert_failure` (for cases where the stacktrace is not available, or not stored in logs, etc).  Unfortunately, this exception constructor takes a tuple argument, and extending it could break some code (do people explicitly actually pattern-match on it, in practice?).  One could always add the symbolic path to the filename component of the payload, but it's a bit ugly.  Switching to an (inline?) record would be better, if we can convince ourselves the amount of code breakage is small enough.

alainfrisch commented:
> A new scope is introduced by a module binding or a module-level let binding (but not an expression-level let ... in ... binding)

What's the argument for keeping local module binding names but not local value names?  If the location is within a local function (or even within a non-functional local binding) with a clear name, it seems useful to keep that name in the path.

stedolan commented:
@lpw25
> I found that the nicest "name" for a function was along the lines of: [...]

This sounds like a solution to a slightly different problem.

In an inlining report you need to be able to refer to particular functions, since inlining is something that happens to functions. The problem being solved is "name this function", and writing `M.foo` for two separate functions (the one bound to `M.foo` and an anonymous function inside it) would indeed be confusing.

In a backtrace, the problem being solved is "describe this location", where locations in question are not functions but callsites (or raises, etc.). In particular, text like `the anonymous function at foo.ml:3,1-10` is wrong: there's no function at the described location.

The goal here is to describe locations approximately. Half the point is that multiple nearby locations get the same description. I'm open to suggestions about what should contribute to a description (classes should definitely be in, as @alainfrisch points out that it's weird to include local modules but not local functions, and I'm still on the fence about marking anonymous functions). But I think it's important to realise that this is a different problem from naming functions for an inlining report.

lpw25 commented:
That difference is why I suggested that the full description might be too detailed.

> In particular, text like `the anonymous function at foo.ml:3,1-10` is wrong: there's no function at the described location.

To clarify, the full backtrace slot with the full description would be something like:

    Called from the anonymous function at foo.ml:3,1-10 within Foo.bar in file "list.ml", line 110, characters 12-15

i.e. the location I was mentioning above was an additional location, not to be confused with the location of the call site.

> I'm open to suggestions about what should contribute to a description

I think the distinction between:

    Called from Typecore.type_expect in file "typing/typecore.ml", line 2327, characters 4-156

and

    Called from an anonymous function within Typecore.type_expect in file "typing/typecore.ml", line 2327, characters 4-156

is the most important thing to distinguish.

stedolan committed:
Print function names (derived from Lambda.scoped_location) in backtraces

Function names now appear in backtraces and are available via Printexc.

stedolan committed:
Update testsuite with new backtrace format, and add test for names

stedolan commented:
Update: This patch got rewritten, which made it both longer and less hacky.

Instead of adding Levent nodes to Lambda which get converted to debuginfo separately by each of the backends (native, bytecode, flambda), the updated PR replaces the Location.t type in Lambda with a new `scoped_location` (which is a pair of a Location.t and some information about scopes). This means the scoping information is only computed in one place.

To the reviewer: read this commit-by-commit. Commits 1 and 4 are large and boring (1 threads a scopes parameter through transl{obj,class,mod} and 4 updates tests), while 2 and 3 are small and interesting (2 is the backend support for printing names, 3 specifies how the names are generated)

stedolan commented:
ping?

stedolan committed:
Bootstrap

stedolan committed:
Introduce the Lambda.scoped_location type

stedolan committed:
Replace Location.t with Lambda.scoped_location in Lambda code

This commit threads scopes through translation from Typedtree to
Lambda, extending the scopes when entering functions, modules,
classes and methods.

shindere commented:
I'm not able to review the code, sorry about that. I just wanted to show
appreciation for the work accomplished here and all the efforts put in
making it understandable even to a non-expert like me.

Not sure this will receive enough reviews to be ready before the 4.11
freeze, but hoping for the best.

avsm commented:
Developer meeting discussion: @gasche: is there consensus on the interface here? @lpw25 / @stedolan both agree on the interface now (but PR description has not been updated).  Decision: consider this for merge into 4.11.

gasche commented:
I think the most practical thing to do for this PR is to accept to merge it after the branch/freeze date, once it is reviewed. I'm not in favor of forcing people to work at full speed before next Monday, because this risks decreasing the quality of the result, increasing stress, and I don't see any advantage.

lpw25 commented on lambda/debuginfo.mli:
```diff
@@ -42,14 +43,12 @@ val is_none : t -> bool
 
 val concat: t -> t -> t
 
-val inline: Location.t -> t -> t
```
Would you mind keeping this function? It will make it easier to interpret it more interestingly that "concat" later.

stedolan replied:
This function was useless - it was exactly `inline a b = concat (from_location a) b`, only used at a point where the code should have already converted `Location.t` into `Debuginfo.t` but still hadn't.

I'm instead interpreting your request as "`inline` is more descriptive name than `concat` for the function above" and making the corresponding change.

lpw25 commented on middle_end/flambda/closure_conversion.ml:
```diff
@@ -207,7 +207,7 @@ let rec close t env (lam : Lambda.lambda) : Flambda.t =
   | Lfunction { kind; params; body; attr; loc; } ->
-    let name = Names.anon_fn_with_loc loc in
+    let name = Names.anon_fn_with_loc (Lambda.raw_location loc) in
     let closure_bound_var = Variable.create name in
```
Since its the only use of this function, and any other uses would need to be similarly adjusted, you should probably push this conversion into the definition of `anon_fn_with_loc`.

lpw25 commented on asmcomp/cmm_helpers.ml:
```diff
@@ -1470,7 +1470,7 @@ struct
   let make_switch loc arg cases actions =
-    let dbg = Debuginfo.from_location loc in
+    let dbg = Debuginfo.from_location (Lambda.of_raw_location loc) in
     let actions = Array.map (fun expr -> expr, dbg) actions in
```
We could avoid going backwards and forwards between representations here and a couple of other places by just making `Switch.Make` polymorphic in the location type it uses.

stedolan replied:
That worked nicely, thanks. Now the Cmm uses of Switch use Debuginfo.t, which is the appropriate debug info type for Cmm, rather than either locations or scoped locations.

stedolan commented:
>  (but PR description has not been updated)

This is done now.

gasche commented:
For me the missing piece was that @lpw25 is happy with the new messages that are proposed. He said so during the meeting, but there is no sign of this in the PR that I can find. I shamefully admit that this decreased my own motivation/effectiveness to do a review: it's a lot of work to get into these low-level details, and it's less exciting to do so when one has the feeling that the PR may linger forever because people disagree (for reasonable reasons) about how to print the result. 

P.S.: thinking about this again, you mentioned during the meeting that you have tools that depend on this PR. Again, maybe this could have been emphasized in the PR discussion, to help prioritize review work. (Also, are the tools available somewhere for the curious?)

lpw25 commented on lambda/debuginfo.ml:
```diff
@@ -25,6 +25,7 @@ type item = {
   dinfo_start_bol: int;
   dinfo_end_bol: int;
   dinfo_end_line: int;
+  dinfo_scopes: Lambda.lambda_scopes;
```
Could you reverse this dependency and put the "lambda_scopes" in here? The scopes don't seem to be lambda-specific -- they last all the way through to emit -- so it makes more sense for them to be part of debuginfo.

lpw25 commented on lambda/lambda.ml:
```diff
@@ -271,17 +271,69 @@ type function_attribute = {
+  | Ls_module_definition of Ident.t
+  | Ls_class_definition of Ident.t
+  | Ls_method_definition of Asttypes.label loc
+
```
Is there a particular reason to use `Ident.t` rather than `string`? As far as I can see the stamp part has lost its meaning here and, for example, would not be updated by a substitution or during inlining.

lpw25 commented on lambda/lambda.ml:
```diff
@@ -271,17 +271,69 @@ type function_attribute = {
+       (match name.[0] with
+        | 'a'..'z' | 'A'..'Z' | '_' | '0'..'9' -> name
+        | _ -> "(" ^ name ^ ")")
+  | Ls_method_definition name ->
```
Could you lift this logic into its own function with a descriptive name? It would be easier to read the name than work out what was happening from the code. I also suspect this logic appears elsewhere in the compiler and might be shared.

lpw25 commented on lambda/translcore.ml:
```diff
@@ -57,11 +58,11 @@ let transl_extension_constructor env path ext =
       Lprim (Pmakeblock (Obj.object_tag, Immutable, None),
-        [Lconst (Const_base (Const_string (name, loc, None)));
+        [Lconst (Const_base (Const_string (name, raw_location loc, None)));
          Lprim (prim_fresh_oo_id, [Lconst (Const_base (Const_int 0))], loc)],
```
Should probably avoid going back and forth here.

lpw25 commented on lambda/translcore.ml:
```diff
@@ -172,20 +173,21 @@ let event_function exp lam =
            Lconst(Const_block(0,
-              [Const_base(Const_string (fname, loc, None));
+              [Const_base(Const_string (fname, raw_location loc, None));
                Const_base(Const_int line);
```
Going back and forth unnecessarily

lpw25 commented on lambda/translcore.ml:
```diff
@@ -758,31 +785,65 @@ and transl_function loc return untuplify_fn repr partial (param:Ident.t) cases =
+   Goes to some trouble to avoid introducing many new anonymous function
+   scopes, as `let f a b = ...` is desugared to several Pexp_fun *)
+and transl_scoped_exp ~scopes ?scope expr =
+  match scope, expr.exp_desc with
```
I don't think the `scope` parameter should be optional. It is explicitly passed at all call sites, and there is already `transl_exp` for places where it isn't needed. It should probably also be called `scope_item`.

stedolan replied:
Good idea. Making this change suggested some other refactorings, so the interface is simpler now.

lpw25 commented on lambda/translcore.ml:
```diff
@@ -758,31 +785,65 @@ and transl_function loc return untuplify_fn repr partial (param:Ident.t) cases =
 *)
-and transl_let rec_flag pat_expr_list =
+and transl_let ~scopes ?(in_structure=false) rec_flag pat_expr_list =
+  let transl_bound_exp pat expr =
```
Maybe lift this function out. It disrupts the flow a bit here.

lpw25 commented on lambda/lambda.mli:
```diff
@@ -252,19 +252,39 @@ type function_attribute = {
+  | Ls_method_definition of Asttypes.label loc
+
+type lambda_scopes = lambda_scope_item list
+
```
I think this type could use some smart constructors, in particular:
```ocaml
val enter_module_definition : lambda_scopes -> Ident.t -> lambda_scopes
val enter_module_definition_opt : lambda_scopes -> Ident.t option -> lambda_scopes
val enter_class_definition : lambda_scopes -> Ident.t -> lambda_scopes
```

lpw25 commented on lambda/lambda.mli:
```diff
@@ -252,19 +252,39 @@ type function_attribute = {
+
+val raw_location : scoped_location -> Location.t
+val of_raw_location : ?scopes:lambda_scopes -> Location.t -> scoped_location
+val string_of_scope_item : lambda_scope_item -> string
```
I think that the `scopes` parameter should not be optional. It should not really be omitted within the compiler, leaving only a couple of places outside the compiler where an empty scopes parameter can easily be passed explicitly.

lpw25 commented on lambda/matching.ml:
```diff
@@ -1717,10 +1717,12 @@ let make_constr_matching p def ctx = function
           | Cstr_block _ ->
-              make_field_args p.pat_loc Alias arg 0 (cstr.cstr_arity - 1) argl
+              make_field_args (of_raw_location p.pat_loc)
+                Alias arg 0 (cstr.cstr_arity - 1) argl
```
It seems a shame to give all locations scopes except for those coming from patterns. Wiring the `scopes` parameter though the pattern matching code seems particularly easy since it won't change at all.

I had thought that not doing so would also mean giving an inferior backtrace for code like:
```ocaml
let l = lazy (raise Not_found)

let () =
  let (lazy ()) = l in ()
```
however it seems the backtrace is already missing the relevant frames for that code. I intend to write a PR to fix that, at which point it would be good to have scopes available in the pattern matching code.

stedolan replied:
> Wiring the scopes parameter though the pattern matching code seems particularly easy since it won't change at all.

The amount of work involved here has nothing to do with how much it changes!

lpw25 commented on lambda/translprim.ml:
```diff
@@ -787,7 +788,8 @@ let primitive_needs_event_after = function
 let transl_primitive_application loc p env ty path exp args arg_exps =
-  let prim = lookup_primitive_and_mark_used loc p env (Some path) in
+  let prim = lookup_primitive_and_mark_used
+               (raw_location loc) p env (Some path) in
```
I think this indentation is too weird. (I really dislike leaving `in` on the end of the line like that, but that style at least is already present in the code base. Whereas having a `let` definition cross multiple lines without also breaking the line after the `=` is I think just against the unwritten house style).

lpw25 commented on bytecomp/instruct.mli:
```diff
@@ -37,12 +37,13 @@ type compilation_env =
     ev_loc: Location.t;                 (* Location in source file *)
     ev_kind: debug_event_kind;          (* Before/after event *)
+    ev_defname: string;                 (* Enclosing definition *)
     ev_info: debug_event_info;          (* Extra information *)
```
It doesn't need to be in this PR, but it might be worth considering how this information could be used in ocamldebug. For example, I think updating the code in `Show_information.show_one_frame` would allow changing the result of ocamldebug's `backtrace` command to match the new backtrace printing.

lpw25 reviewed (commented):
I've reviewed the compiler parts up to cmm and they look pretty good. A few parts could be tidied up. In particular, I think that the scope types should drop the `lambda_` from their name and move to `Debuginfo`, and I would like the locations generated in `matching.ml` to also have scopes.

gasche commented:
Thanks @lpw25 for your review; I like the style of your review comments, some I could have written myself. (Yay for complaining about optional arguments.) Let me know I can help for the Matching/Switch part. I worked on having more locations in Switch in <https://github.com/ocaml/ocaml/compare/trunk...gasche:switch-loc> when reviewing #2294. There the purpose was to have more precise locations for the tests; I'm not sure you need this here (I haven't looked at the code), @lpw25 suggests that just parametrization is enough.

mshinwell commented:
@lpw25 I will read the backend parts, probably today -- can you confirm that I need to read the runtime changes, Cmm_helpers and Emitaux?

lpw25 commented:
@mshinwell I think those are all the parts I didn't check

lpw25 commented:
Looking at the names printed in tests/backtrace/names.ml there are a couple of bits I might change:
- I might use `#` as the separator between class and method names. It feels more natural to me since that is what is used for method calls.
- I would prefer to use something like `/` for things underneath value names, reserving `.` for normal paths. However, I know Stephen doesn't agree and I don't think it matters that much, so I don't expect this to change.
- I would use `object` rather than `class` for object literals. Although I don't see where that comes from in the code, so maybe that is not easy to change.

stedolan committed:
Changes

stedolan commented:
Thanks for the review! I've changed to using `#`, and the `class -> object` change turned out to be quite easy. (As you guessed, I'm not planning to use `/` for values - it's unfamiliar, and as explained in the PR description these names don't always correspond to normal paths for many reasons).

I've pushed the changes after review as one massive commit. After you've read them, I'll split up and merge them into the relevant commits here, to keep history readable.

stedolan commented:
I am unsure which part of `Changes` this change should live in. I've guessed "Runtime system", but I'm open to other suggestions.

gasche commented:
Yes, "Runtime system" sounds right.

I just looked at the new reference files for backtraces. One thing that looks a bit redundant is that they all say

`raised at Foo.bar in file "path/foo.ml" at <source location>`

Wouldn't it make sense ot elide the `Foo.` in the path-like part, given that this information is already in the file path? (In the case where the path is just `Foo`, one could say "called from the toplevel in file ..." or just "called from file...".)

While we are in the land of silly interface questions, personally I tend to agree with Leo it is awkward to use a path-like notations for things that are not path (notably, it uses `a.b` in cases where `a` does not contain `b` as component, just as an internal sub-declaration, which I find disturbing). I would also prefer `/` or `>` or `->`.

mshinwell commented:
I've been using this patch on the Flambda 2.0 branch and have found that the qualified module name is my primary reference -- I don't really look at the filename any more.  I think it would be worse to have to look in two disparate places on the same line, so I'd prefer the format to remain as-is.

mshinwell commented on stdlib/printexc.mli:
```diff
@@ -269,6 +269,16 @@ module Slot : sig
+  (** [name slot] returns the name of the function or definition
+      enclosing the location referred to by the slot.
+
+      [name slot] returns None if the name is  unavailable, which
```
Double space before "unavailable"

mshinwell commented on stdlib/printexc.ml:
```diff
@@ -143,8 +144,8 @@ let format_backtrace_slot pos slot =
   | Known_location l ->
-      Some (sprintf "%s file \"%s\"%s, line %d, characters %d-%d"
-              (info l.is_raise) l.filename
+      Some (sprintf "%s %s in file \"%s\"%s, line %d, characters %d-%d"
```
This looks like it will emit a double space if the name is empty.  Same comment for the version in `runtime`.

stedolan replied:
Fixed by ensuring names are never empty.

mshinwell commented on bytecomp/bytegen.ml:
```diff
@@ -306,10 +302,14 @@ let add_event ev =
       to prevent the debugger to stop at every single allocation. *)
 let add_pseudo_event loc modname c =
   if !Clflags.debug then
+    let ev_defname = match loc with
```
I suggest factoring out these three lines, from here and below.

mshinwell replied:
Also, could we have the `match` on a new line?

mshinwell reviewed (approved):
I read `asmcomp`, `bytecomp`, `runtime` and `stdlib` changes as I was instructed and I approve these changes (to be clear, just the ones in these directories).  There are a couple of minor comments here whose fixes I don't need to re-approve.

gasche commented:
> I [..] have found that the qualified module name is my primary reference -- I don't really look at the filename any more. I think it would be worse to have to look in two disparate places on the same line, so I'd prefer the format to remain as-is.

Maybe the problem is that the file name is printed second, while being more global than the raise scope. Instead of the current 

```
Called from Backtrace.f in file "backtrace.ml", line 12, characters 42-53
Called from Dynlink.Native.run.(fun) in file "otherlibs/dynlink/native/dynlink.ml", line 85, characters 12-29
```

we could use

```
Called from file "backtrace.ml" in f: line 12, characters 42-53
Called from file "otherlibs/dynlink/native/dynlink.ml" in Native.run.(fun): line 85, characters 12-29
```

Then reports are clearly ordered by precision: first the path to the source file, then the scope within the source file, then the source location within the scope. (And again the redundancy between filename and compilation unit name is gone.)

(The use of `:` after the scope path is designed to make it easy to identify the source location visually, and also when grepping in editor modes.)

shindere commented:
Gabriel Scherer (2020/04/21 08:13 -0700):
> Maybe the problem is that the file name is printed second, while being more global than the raise scope. Instead of the current
>
> ```
> Called from Backtrace.f in file "backtrace.ml", line 12, characters 42-53
> Called from Dynlink.Native.run.(fun) in file "otherlibs/dynlink/native/dynlink.ml", line 85, characters 12-29
> ```
>
> we could use
>
> ```
> Called from file "backtrace.ml" in f: line 12, characters 42-53
> Called from file "otherlibs/dynlink/native/dynlink.ml" in Native.run.(fun): line 85, characters 12-29
> ```

That looks better to me, too.

damiendoligez commented:
I think, like @mshinwell, that it's better to keep the module name. Having a relative path (relative to a different thing in each line) just seems too weird.

stedolan commented:
The current format is:
```
Called from <thing> in <location>
```
When reading a backtrace, I scan down the list until I see an interesting `<thing>`, and then continue reading that line to learn its `<location>`, which I then load in my editor.

Both changes proposed (splitting `<location>` in two and putting it either side of `<thing>`, and omitting the part of `<thing>` which can be deduced from `<location>`) make this harder to do.

stedolan commented:
The current format is:
```
Called from <thing> in <location>
```
When reading a backtrace, I scan down the list until I see an interesting `<thing>`, and then continue reading that line to learn its `<location>`, which I then load in my editor.

Both changes proposed (splitting `<location>` in two and putting it either side of `<thing>`, and omitting the part of `<thing>` which can be deduced from `<location>`) make this harder to do.

stedolan commented:
The current format is:
```
Called from <thing> in <location>
```
When reading a backtrace, I scan down the list until I see an interesting `<thing>`, and then continue reading that line to learn its `<location>`, which I then load in my editor.

Both changes proposed (splitting `<location>` in two and putting it either side of `<thing>`, and omitting the part of `<thing>` which can be deduced from `<location>`) make this harder to do.

stedolan commented:
The current format is:
```
Called from <thing> in <location>
```
When reading a backtrace, I scan down the list until I see an interesting `<thing>`, and then continue reading that line to learn its `<location>`, which I then load in my editor.

Both changes proposed (splitting `<location>` in two and putting it either side of `<thing>`, and omitting the part of `<thing>` which can be deduced from `<location>`) make this harder to do.

stedolan commented:
The current format is:
```
Called from <thing> in <location>
```
When reading a backtrace, I scan down the list until I see an interesting `<thing>`, and then continue reading that line to learn its `<location>`, which I then load in my editor.

Both changes proposed (splitting `<location>` in two and putting it either side of `<thing>`, and omitting the part of `<thing>` which can be deduced from `<location>`) make this harder to do.

mshinwell commented:
Even ignoring the "relative" issue, I think this is worse:
```
Called from file "backtrace.ml" in f: line 12, characters 42-53
```
because the position on the line of the thing I'm primarily searching for (the function/module name) is different for each line.

I do wonder, though, whether we should make these backtraces more concise like other tools such as `gdb` (given that we have the opportunity).  For example:
```
Called from Backtrace.f at backtrace.ml:12,42-53
```
This would help with the one problem I have found with the new messages, namely that they can easily run off the right-hand side of the terminal.

lpw25 reviewed (approved):
The code LGTM now

gasche commented:
Okay, let's go with the interface that people that have played with the stuff in practice like better.

stedolan commented:
> I do wonder, though, whether we should make these backtraces more concise like other tools such as `gdb` (given that we have the opportunity). For example:
> 
> ```
> Called from Backtrace.f at backtrace.ml:12,42-53
> ```

I also prefer this more standard format - it's shorter and I find it easier to read.

I did worry that making this change would break programs that parse backtraces. I grepped around on OPAM and found only two instances of this - one in `expect_test_helpers_kernel` and one in emacs config files. Happily, in both of these instances there was already another regex present for the standard gnu format.

However, I did run into about a dozen parsers for the slightly different format that the compiler produces in error messages (it differs from backtraces by using a capital `F` in `File`). Are you proposing to change this format as well?

FWIW, I'm working on some Memprof-based tools at https://github.com/janestreet/memtrace (as @gasche mentioned above), and I've been using the gnu format to refer to locations. I've found it easier to read than `file ..., line ..., characters ...`. ([Example output here](https://gist.github.com/stedolan/48514431abeff6a769d9d213c86f8bc0)).

stedolan commented:
I've just rewritten history to make the commit log a bit more readable, but the code hasn't changed since review.

Thinking about it some more, I'd like to leave the discussion of `file ..., line...` vs. `file:line` to a different PR. It's a discussion I'd like to have, but it may take a while and I'd like avoid delaying merging this patch. (This is a large if mechanical change, and I've already had to rebase it too many times)

mshinwell commented:
This seems fine and I think this should be ready for merging now.  However I would prefer if we could reach a consensus on the backtrace format before 4.11.  There is some amount of work involved each time the backtrace format changes.

gasche commented:
Note that if you don't want to write tools that depend on the backtrace format, you can write your own backtrace printer (generating JSON, for example), and have your tools consume that.

damiendoligez commented:
I vote for the GNU format, and I dream of using it in the compiler's error messages too.

About the uppercase F, I seem to remember a time when it was needed for emacs to recognize error messages in compiler output.

shindere commented:
Damien Doligez (2020/04/22 14:58 +0000):
> I vote for the GNU format, and I dream of using it in the compiler's
> error messages too.

So do I.

damiendoligez commented:
Note: the feature freeze for 4.11 has already happened, but at the developer meeting a few days ago we decided that this PR gets a special pass to be cherry-picked to the 4.11 branch.

stedolan commented:
Speaking of which: I've just rebased and rebootstrapped again. I'm going to merge and cherry pick to 4.11 as soon as CI passes.
